sklearn/datasets/openml.py

```diff
@@ -77,20 +77,67 @@ def _open_openml_url(openml_path, data_home):
 
     Returns
     -------
-    result : stream
-        A stream to the OpenML resource
+    result : bytes
+        Byte content of resource
     """
     def is_gzip(_fsrc):
         return _fsrc.info().get('Content-Encoding', '') == 'gzip'
 
     req = Request(_OPENML_PREFIX + openml_path)
     req.add_header('Accept-encoding', 'gzip')
 
+    def _md5_validated_bytestream(fsrc, expected_md5=None, chunk_size=512):
+        """
+        Takes in a byte-stream, reads in chunks and returns bytes.
+        If expected_md5 is not none, keeps md5 checksum state while streaming
+        and validates post stream consumption.
+
+        Parameters
+        ----------
+        fsrc : io.BufferedIOBase
+            input stream to read bytes from
+
+        expected_md5 : str
+            expected md5-checksum value
+
+        chunk_size : int
+            size of chunks to read at a time from stream
+
+        Returns
+        -------
+        fsrc_bytes : bytes
+            equivalent to fsrc_bytes.read() but with md5 validation if
+            expected_md5 is provided
+
+        Raises
+        ------
+        ValueError :
+            if expected_md5 does not match actual md5-checksum of stream
+        """
+        fsrc_bytes = bytes()
+        file_md5 = hashlib.md5() if expected_md5_checksum else None
+        while True:
+            data = fsrc.read(chunk_size)
+            if not data:
+                break
+            if expected_md5:
+                file_md5.update(data)
+            fsrc_bytes += data
```

Comment:
I am concerned that every assignment here will create a new `bytes` since `bytes` is immutable. Would `bytearray` and `bytearray.extend` be better in this case?